todo_app/app.py

```diff
@@ -25,64 +48,96 @@ def index():
-        
-        toggle_item = request.form.get('delete_id')
-
-        for card in allcards:
-            if toggle_item == str(card.id):
-                delete_todo_card(card.id)
+        if 'LOGIN_DISABLED' in app.config or current_user.role == 'writer':
+            allcards = get_todo_cards()
+            lists = {'todo','doing','done'}
+            
+            toggle_item = request.form.get('delete_id')
+
+            for card in allcards:
+                if toggle_item == str(card.id):
+                    delete_todo_card(card.id)
         
         return redirect(request.headers.get('Referer'))
 
 
     @app.route('/Doing', methods=['POST'])
+    @login_required
     def in_progress():
-        allcards = get_todo_cards()
-        lists = {'todo','doing','done'}
+        if 'LOGIN_DISABLED' in app.config or current_user.role == 'writer':
+            allcards = get_todo_cards()
+            lists = {'todo','doing','done'}
 
-        toggle_item = request.form.get('item_id')
+            toggle_item = request.form.get('item_id')
 
-        for card in allcards:
-            if toggle_item == str(card.id):
-                move_todo_card(card.id, 'doing')
+            for card in allcards:
+                if toggle_item == str(card.id):
+                    move_todo_card(card.id, 'doing')
         
         return redirect(request.headers.get('Referer'))
 
     @app.route('/Done', methods=['POST'])
+    @login_required
     def Done():
-        allcards = get_todo_cards()
-        lists = {'todo','doing','done'}
+        if 'LOGIN_DISABLED' in app.config or current_user.role == 'writer':
+            allcards = get_todo_cards()
+            lists = {'todo','doing','done'}
 
-        toggle_item = request.form.get('item_id')
+            toggle_item = request.form.get('item_id')
 
-        for card in allcards:
-            if toggle_item == str(card.id):
-                move_todo_card(card.id, 'done')
+            for card in allcards:
+                if toggle_item == str(card.id):
+                    move_todo_card(card.id, 'done')
         return redirect(request.headers.get('Referer'))
 
+    @app.route('/login')
+    def login_callback():
+        callback_code = request.args.get("code")
+        github_client =  WebApplicationClient(os.environ.get('CLIENTID'))
+        github_token = github_client.prepare_token_request("https://github.com/login/oauth/access_token", code=callback_code) 
+        github_access = requests.post(github_token[0], headers=github_token[1], data=github_token[2], auth=(os.environ.get('CLIENTID'), os.environ.get('CLIENTSECRET')))
+        github_json = github_client.parse_request_body_response(github_access.text)
+        github_user_request_param = github_client.add_token("https://api.github.com/user")
+        github_user = requests.get(github_user_request_param[0], headers=github_user_request_param[1]).json()
+        
+        login_user(User(github_user))
```

Comment:
I missed this last time as the authorisation wasn't hiding anything, but I don't think your app ever treats anyone as a user - the "id" field passed through here (`github_user`) is actually a large json object with lots of fields, so the below never passes
```python
if self.id == 'JoeBawn':
            return 'writer'
```

Instead, we want to just read a specific field out of this response;
```suggestion
        login_user(User(github_user['login']))
```